Replace Arsenal Crates with Sigils

Galaco opened the pull request:
## Overview

Sigils can be used as arsenal crates, removed arsenal crates and packs from worth and shop, allow purchase anywhere in Objective maps.

Resolves #6

## Demo

![Screenshot 2020-09-15 at 14 52 14](https://user-images.githubusercontent.com/3932269/93172995-de81e100-f766-11ea-9cce-005d3bf7ee9c.png)

## Notes


I didn't increase the distance you have to be from the sigil to buy; it uses the same arsenal crate value, which requires close proximity to the sigil. 

Arsenal crates/packs are still supported b/c legacy, but have been removed from worth and shop menus.

## Testing Instructions

* Join the server as a human
* Look at a sigil after Wave 1 starts
* Hint should appear when within purchase range
* Press F2 to use the buy menu

## Checklist

- [x] `fixup!` commits have been squashed
- [ ] README.md updated if necessary to reflect the changes

Galaco commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/sh_options.lua:
```diff
@@ -187,8 +187,6 @@ item.SkillRequirement = SKILL_U_STRENGTHSHOT
 item =
 GM:AddStartingItem("antidoteshot",		ITEMCAT_TOOLS,			40,				"weapon_zs_antidoteshot")
 item.SkillRequirement = SKILL_U_ANTITODESHOT
-GM:AddStartingItem("arscrate",			ITEMCAT_DEPLOYABLES,			50,				"weapon_zs_arsenalcrate")
```
Removed Arsenal Crates and Arsenal packs from worth and shop

Galaco commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/obj_player_extend.lua:
```diff
@@ -230,14 +230,21 @@ function meta:MeleeViewPunch(damage)
 
 function meta:NearArsenalCrate()
+	-- Always near an arsenal crate in objective maps
+	if GAMEMODE.ObjectiveMap then
```
This continues to support obj since we've removed the ability to buy stuff after round begins.

Galaco commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/init.lua:
```diff
@@ -3913,31 +3913,6 @@ function GM:WaveStateChanged(newstate)
 					gamemode.Call("GiveDefaultOrRandomEquipment", pl)
 				end
 			end
-
-			-- We should spawn a crate in a random spawn point if no one has any.
-			if not self.ZombieEscape and #ents.FindByClass("prop_arsenalcrate") == 0 then
-				local have = false
-				for _, pl in pairs(humans) do
-					if pl:HasWeapon("weapon_zs_arsenalcrate") then
-						have = true
-						break
-					end
-				end
-
-				if not have and #humans >= 1 then
-					local spawn = self:PlayerSelectSpawn(humans[math.random(#humans)])
-					if spawn and spawn:IsValid() then
-						local ent = ents.Create("prop_arsenalcrate")
-						if ent:IsValid() then
-							ent:SetPos(spawn:GetPos() + Vector(0, 0, 8))
-							ent:Spawn()
-							ent:DropToFloor()
-							ent:SetCollisionGroup(COLLISION_GROUP_DEBRIS_TRIGGER) -- Just so no one gets stuck in it.
-							ent.NoTakeOwnership = true
-						end
-					end
-				end
-			end
```
Not necessary anymore now that arsenals cannot be purchased and sigils will always exist.

rbreslow reviewed (approved):
Nice job. I'm going crazy trying to think of any other edge case that this could effect, but I think we're OK.

If you want, you could also update `cl_targetid.lua`.

```diff
diff --git a/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua b/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
index 3d468c1..255068e 100644
--- a/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
+++ b/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
@@ -106,9 +106,11 @@ function GM:DrawSigilTargetHint(ent, fade)
        util.ColorCopy(color_white, colTemp)

        draw.SimpleTextBlur("Sigil", "ZSHUDFontSmaller", x, y, colTemp, TEXT_ALIGN_CENTER)
-       y = y + draw.GetFontHeight("ZSHUDFontSmaller") + 0
+       local fontHeight = draw.GetFontHeight("ZSHUDFontSmaller")
+       y = y + fontHeight + 0

        draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
+       draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight / 2, colTemp, TEXT_ALIGN_CENTER)
 end

 GM.TraceTarget = NULL
```

Also, same comment on my last PR, I'd update your PR body so that it resolves the current issue.

Galaco commented:
> Nice job. I'm going crazy trying to think of any other edge case that this could effect, but I think we're OK.
> 
> If you want, you could also update `cl_targetid.lua`.
> 
> ```diff
> diff --git a/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua b/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
> index 3d468c1..255068e 100644
> --- a/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
> +++ b/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua
> @@ -106,9 +106,11 @@ function GM:DrawSigilTargetHint(ent, fade)
>         util.ColorCopy(color_white, colTemp)
> 
>         draw.SimpleTextBlur("Sigil", "ZSHUDFontSmaller", x, y, colTemp, TEXT_ALIGN_CENTER)
> -       y = y + draw.GetFontHeight("ZSHUDFontSmaller") + 0
> +       local fontHeight = draw.GetFontHeight("ZSHUDFontSmaller")
> +       y = y + fontHeight + 0
> 
>         draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
> +       draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight / 2, colTemp, TEXT_ALIGN_CENTER)
>  end
> 
>  GM.TraceTarget = NULL
> ```
> 
> Also, same comment on my last PR, I'd update your PR body so that it resolves the current issue.

This is a good idea; I'd rather add this here if I can, than come back to it later. Will hold off merging for now.

EDIT: This is actually great, I lost count of how often I saw new players asking how to buy items; this will add a huge amount of visibility on how to buy items.

Galaco committed:
Replace Arsenal Crates with Sigils

Sigils can be used as arsenal crates, removed arsenal crates and packs from worth and shop, allow purchase anywhere in Objective maps

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,14 @@ function GM:DrawSigilTargetHint(ent, fade)
 	draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
+	
+	if self.GetWave() > 0 then
+    	draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight / 2, colTemp, TEXT_ALIGN_CENTER)
```
```suggestion
    	    draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight / 2, colTemp, TEXT_ALIGN_CENTER)
```

rbreslow replied:
I'd also consider doing something like this outside of the render hook. This translation has an exclamation point since it was designed for drawing in the center of the HUD, while the string above it does not.

```lua
local press_f2_for_the_points_shop = translate.Get("press_f2_for_the_points_shop"):gsub("%!", "")
```

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,14 @@ function GM:DrawSigilTargetHint(ent, fade)
 
 	draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
+	
+	if self.GetWave() > 0 then
```
I would consider applying a check that uses the `GM:PlayerCanPurchase` hook.

https://github.com/rbreslow/zs/blob/00f1a80f382d72ceb91912941587e67b5b735e05/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_init.lua#L911-L913

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,14 @@ function GM:DrawSigilTargetHint(ent, fade)
 	draw.SimpleTextBlur("Sigil", "ZSHUDFontSmaller", x, y, colTemp, TEXT_ALIGN_CENTER)
-	y = y + draw.GetFontHeight("ZSHUDFontSmaller") + 0
+	local fontHeight = draw.GetFontHeight("ZSHUDFontSmaller")
+	y = y + fontHeight + 0
```
`+ 0`?

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,14 @@ function GM:DrawSigilTargetHint(ent, fade)
 
 	draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
+
+	if MySelf:IsValid() and MySelf:Team() == TEAM_HUMAN and self:PlayerCanPurchase(MySelf) then
```
The first conditions are unnecessary since `GM:DrawSigilTargetHint` is only invoked for humans:
 https://github.com/rbreslow/zs/blob/cb6206b43c8ea4896305cf26979de75a837fce23/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua#L149-L151

Additionally, I don't think it's necessary to check `MySelf:IsValid()` because order rendering hooks seem to think it's fine. e.g.

https://github.com/rbreslow/zs/blob/00f1a80f382d72ceb91912941587e67b5b735e05/src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_init.lua#L853-L858

`MySelf` is defined in `GM:InitPostEntity` so there won't be many cases where it's invalid.

Galaco replied:
You're right; these checks have been removed.

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,14 @@ function GM:DrawSigilTargetHint(ent, fade)
 	draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y, colTemp, TEXT_ALIGN_CENTER)
+
+	if MySelf:IsValid() and MySelf:Team() == TEAM_HUMAN and self:PlayerCanPurchase(MySelf) then
+    	draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"):gsub("%!", ""), "ZSHUDFontTiny", x, y + fontHeight / 2, colTemp, TEXT_ALIGN_CENTER)
```
Be mindful that you're performing pattern matching in a render hook. This could be expensive. Fixing it may be a premature optimization, but that's why I declared this once outside of scope in my example.

The downside though is if a user switches their language, we may not have the up-to-date translation. 

Also, you are calculating the y-value math inline with the function here, but you're doing it separately for the first draw call.

Galaco replied:
Removed the pattern matching. Leaving the `!` prefix doesn't affect how this text is rendered (see screenshot in PR description). Translations therefore wont be an issue here.

rbreslow replied:
> Leaving the `!` prefix doesn't affect how this text is rendered

I disagree. I think it's a matter of consistency, where 1 message ends in an exclamation and the other doesn't. If you're leaning towards leaving in the inconsistency, I'm fine with that.

rbreslow commented on src/garrysmod/gamemodes/zombiesurvival/gamemode/cl_targetid.lua:
```diff
@@ -106,9 +106,13 @@ function GM:DrawSigilTargetHint(ent, fade)
+	draw.SimpleTextBlur("Press E to teleport", "ZSHUDFontTiny", x, y + fontHeight, colTemp, TEXT_ALIGN_CENTER)
+
+	if self:PlayerCanPurchase(MySelf) then
+    	draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight * 1.5, colTemp, TEXT_ALIGN_CENTER)
```
Nitpick, but there's still an indent missing here. I know we haven't established a code style yet, but I've been trying to do 4 spaces.

```suggestion
    	    draw.SimpleTextBlur(translate.Get("press_f2_for_the_points_shop"), "ZSHUDFontTiny", x, y + fontHeight * 1.5, colTemp, TEXT_ALIGN_CENTER)
```

rbreslow reviewed (approved):
Nice job!

You may merge at your discretion, but please consider my comments first. Also, I realize the PR template used the `Connects` keyword, but you have to update it to `Fixes`, `Resolves`, or `Closes` in order for this PR to close out the associated issue.

Galaco commented:
> Nice job!
> 
> You may merge at your discretion, but please consider my comments first. Also, I realize the PR template used the `Connects` keyword, but you have to update it to `Fixes`, `Resolves`, or `Closes` in order for this PR to close out the associated issue.

Corrected formatting issues.

While I believe your point regarding text inconsistency is valid, I do feel that its out of scope of this issue, but is potentially indicative of a separate task involving a more general review of the language used

rbreslow commented:
> While I believe your point regarding text inconsistency is valid, I do feel that its out of scope of this issue, but is potentially indicative of a separate task involving a more general review of the language used

Good point. I agree.
